src/main/java/org/opensearch/geospatial/ip2geo/action/DeleteDatasourceTransportAction.java

```diff
@@ -0,0 +1,132 @@
+    private final IngestService ingestService;
+    private final DatasourceFacade datasourceFacade;
+
+    /**
+     * Constructor
+     * @param transportService the transport service
+     * @param actionFilters the action filters
+     * @param lockService the lock service
+     * @param ingestService the ingest service
+     * @param datasourceFacade the datasource facade
+     */
+    @Inject
+    public DeleteDatasourceTransportAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final Ip2GeoLockService lockService,
+        final IngestService ingestService,
+        final DatasourceFacade datasourceFacade
+    ) {
+        super(DeleteDatasourceAction.NAME, transportService, actionFilters, DeleteDatasourceRequest::new);
+        this.lockService = lockService;
+        this.ingestService = ingestService;
+        this.datasourceFacade = datasourceFacade;
+    }
+
+    /**
+     * We delete datasource regardless of its state as long as we can acquire a lock
+     *
+     * @param task the task
+     * @param request the request
+     * @param listener the listener
+     */
+    @Override
+    protected void doExecute(final Task task, final DeleteDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        lockService.acquireLock(request.getName(), LOCK_DURATION_IN_SECONDS, ActionListener.wrap(lock -> {
+            if (lock == null) {
+                listener.onFailure(new OpenSearchException("another processor is holing a lock on the resource. Try again later"));
+                return;
+            }
+            try {
+                deleteDatasource(request.getName());
+                listener.onResponse(new AcknowledgedResponse(true));
+            } catch (Exception e) {
+                listener.onFailure(e);
+            } finally {
+                lockService.releaseLock(
+                    lock,
+                    ActionListener.wrap(
+                        released -> { log.info("Released lock for datasource[{}]", request.getName()); },
+                        exception -> { log.error("Failed to release the lock", exception); }
+                    )
+                );
+            }
+        }, exception -> { listener.onFailure(exception); }));
+    }
+
+    private void deleteDatasource(final String datasourceName) throws IOException {
+        Datasource datasource = datasourceFacade.getDatasource(datasourceName);
+        if (datasource == null) {
+            new ResourceNotFoundException("no such datasource exist");
+        }
+
+        setDatasourceStateAsDeleting(datasource);
+        datasourceFacade.deleteDatasource(datasource);
+    }
+
+    private void setDatasourceStateAsDeleting(final Datasource datasource) throws IOException {
+        if (isSafeToDelete(datasource) == false) {
+            new OpenSearchException("datasource is being used by one of processors");
```

Comment:
Why doesnt this have throw before it?

Reply:
before means? A client will see the error message in response when this exception is thrown.

Reply:
When does it get thrown?

Reply:
If a use trying to delete datasource using DELETE api, it check if there is a processor associated with the datasource. If there is, then the exception is thrown and returned back to the user.

Reply:
Talked offline. Missed to add `throw`